src/dac.rs

```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    /// Enable the DAC.
+    pub fn enable(&mut self, apb1: &mut APB1) {
+        match self.id {
+            DacId::One => {
+                apb1.enr().modify(|_, w| w.dac1en().set_bit());
+                self.regs.cr.modify(|_, w| w.en1().set_bit());
+            }
+            DacId::Two => {
+                apb1.enr().modify(|_, w| w.dac2en().set_bit());
+                self.regs.cr.modify(|_, w| w.en2().set_bit());
+            }
+        }
+    }
+
+    /// Disable the DAC
+    pub fn disable(&mut self, apb1: &mut APB1) {
+        match self.id {
+            DacId::One => {
+                self.regs.cr.modify(|_, w| w.en1().clear_bit());
+                apb1.enr().modify(|_, w| w.dac1en().clear_bit());
+            }
+            DacId::Two => {
+                self.regs.cr.modify(|_, w| w.en2().clear_bit());
+                apb1.enr().modify(|_, w| w.dac2en().clear_bit());
+            }
+        }
+    }
+
+    /// Set the DAC value as an integer.
+    pub fn set_value(&mut self, val: u32) {
+        match self.id {
+            DacId::One => match self.bits {
+                DacBits::EightR => self.regs.dhr8r1.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveL => self.regs.dhr12l1.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveR => self.regs.dhr12r1.modify(|_, w| unsafe { w.bits(val) }),
+            },
+            DacId::Two => match self.bits {
+                DacBits::EightR => self.regs.dhr8r2.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveL => self.regs.dhr12l2.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveR => self.regs.dhr12r2.modify(|_, w| unsafe { w.bits(val) }),
+            },
+        }
+    }
+
+    /// Set the DAC voltage. `v` is in Volts.
+    pub fn set_voltage(&mut self, volts: f32) {
+        let val = match self.bits {
+            DacBits::EightR => ((volts / self.vref) * 255.) as u32,
+            DacBits::TwelveL => ((volts / self.vref) * 4_095.) as u32,
+            DacBits::TwelveR => ((volts / self.vref) * 4_095.) as u32,
+        };
+
+        self.set_value(val);
+    }
+
+    // Select and activate a trigger. See f303 Reference manual, section 16.5.4.
+    pub fn set_trigger(&mut self, trigger: Trigger) {
+        match self.id {
+            DacId::One => {
+                self.regs.cr.modify(|_, w| w.ten1().set_bit());
+                self.regs
+                    .cr
+                    .modify(|_, w| unsafe { w.tsel1().bits(trigger.bits()) });
+            }
+            DacId::Two => {
+                self.regs.cr.modify(|_, w| w.ten2().set_bit());
+                self.regs.cr.modify(|_, w| w.tsel2().bits(trigger.bits()));
+            }
+        }
+    }
+
+    /// Independent trigger with single LFSR generation
+    /// See f303 Reference Manual section 16.5.2
+    pub fn trigger_lfsr(&mut self, trigger: Trigger, data: u32) {
+        // todo: This may not be correct.
```

Comment:
I would rather not merge code that might not be correct. Do we have a way to verify this? If not we should maybe just remove it.